### ISSUE TITLE
Secure Storage and Cloud Back-up

### ISSUE THREAD

**Issue by robert-at-hedera** [hashgraph/hedera-wallet-ios#276]
Follow a specific set of requirements on both Android and iOS for a consistent experience and aligned implementations around the storage of the recovery phrase, PIN / biometric authentication information, and their interaction with automatic cloud back-ups.

The wallet app maintains a copy of the recovery phrase for the primary cryptocurrency account.  It also may maintain an authentication PIN if requested by the user (or, in the alternative, some kind of biometric authentication provided by the platform).

These two pieces of information require the highest security.  Storage within the app directory as well as in the cloud are both too insecure to be allowed.  Unfortunately, the most secure location - the platform key store - may be uploaded to the cloud without our app's permission.

The solution is generation of a securely random key on the specific device.  That key is stored in the key store of the platform and protected by the PIN / biometric authentication via platform-level mechanisms.  This key may be automatically uploaded to the cloud.

At the same time, that key is actually used to decrypt the recovery phrase, which is stored in permanent storage in such a way that automatic upload is disallowed by the application.  In this way, access to both the key store and the encrypted recovery phrase file are needed to actually determine the recovery phrase.

**Comment by robert-at-hedera** [comment]
Assigning to Donald for prioritization.

**Comment by robert-at-hedera** [comment]
High priority.  Looking into timeframe.

**Comment by robert-at-hedera** [comment]
I think the realistic timeframe for this is three days, including testing time.  The encryption key and file stuff isn't too bad.  I don't think new keys nor migration are too terrible.  Gracefully handling various scenarios, like after a partial restore or Apple deleting the "please do not delete" file anyway ... testing the edge cases is the expense here.

**Comment by robert-at-hedera** [comment]
I've saved off my work locally and am posting this update as part of refocusing my effort on this task.

So far, I have the RSA 2048 key management and file management.  I can migrate from the old model to the new one.  The basic are done.  The work in progress when I saved this off was working through all the TODOs and FIXMEs that cover all the non-ideal cases.

Going forward, I need to cover the practical missing elements and associated tests.  The final tests are manually covering the primary cases (fresh install, migration, working operations) and then the particular cases for which this issues exists: restoring the keychain and/or app data.

Donald had concerns about how the user experience will change.  He pointed out a scenario that I had not considered: a user who has successfully restored in the past who expects future success with backups.  This user would have to both have the keychain and app data set for upload and restore.

As such, I am going to see if I can get restore to work successfully on my personal test device.  If successful, then Donald and I will discuss how to properly handle messaging users about the change.
